sklearn/utils/stats.py

```diff
@@ -40,33 +45,41 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     # When sample_weight 1D, repeat for each array.shape[1]
     if array.shape != sample_weight.shape and array.shape[0] == sample_weight.shape[0]:
         sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
+
+    # Sort `array` and `sample_weight` along axis=0; set `sample_weight` for nan input
+    # to 0:
     sorted_idx = np.argsort(array, axis=0)
     sorted_weights = np.take_along_axis(sample_weight, sorted_idx, axis=0)
+    sorted_nan_mask = np.take_along_axis(np.isnan(array), sorted_idx, axis=0)
+    sorted_weights[sorted_nan_mask] = 0
 
-    # Find index of median prediction for each sample
+    # Compute the weighted cumulative distribution function (CDF) based on
+    # `sample_weight` and scale `percentile` along it:
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
     adjusted_percentile = percentile / 100 * weight_cdf[-1]
 
-    # For percentile=0, ignore leading observations with sample_weight=0. GH20528
+    # For percentile=0, ignore leading observations with sample_weight=0; see PR #20528
     mask = adjusted_percentile == 0
     adjusted_percentile[mask] = np.nextafter(
         adjusted_percentile[mask], adjusted_percentile[mask] + 1
     )
 
+    # Find index `adjusted_percentile` would have in `weight_cdf`:
     percentile_idx = np.array(
         [
             np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
             for i in range(weight_cdf.shape[1])
         ]
     )
-    percentile_idx = np.array(percentile_idx)
+
     # In rare cases, percentile_idx equals to sorted_idx.shape[0]
     max_idx = sorted_idx.shape[0] - 1
     percentile_idx = np.apply_along_axis(
         lambda x: np.clip(x, 0, max_idx), axis=0, arr=percentile_idx
     )
 
-    col_index = np.arange(array.shape[1])
-    percentile_in_sorted = sorted_idx[percentile_idx, col_index]
-    percentile = array[percentile_in_sorted, col_index]
-    return percentile[0] if n_dim == 1 else percentile
+    col_indices = np.arange(array.shape[1])
+    percentile_in_sorted = sorted_idx[percentile_idx, col_indices]
+    values = array[percentile_in_sorted, col_indices]
+
+    return values[0] if n_dim == 1 else values
```

Comment:
Let's keep the old variable name that was more explicit:

```suggestion
    percentile = array[percentile_in_sorted, col_indices]
    return percentile[0] if n_dim == 1 else percentile
```

or alternatively:


```suggestion
    percentile_value = array[percentile_in_sorted, col_indices]
    return percentile_value[0] if n_dim == 1 else percentile_value
```

or even:

```suggestion
    percentile_score = array[percentile_in_sorted, col_indices]
    return percentile_score[0] if n_dim == 1 else percentile_score
```

to avoid shadowing the input argument. Personally, I am not a big fan of `percentile_score` (because it's not related to what we usually call "score" in the scikit-learn API), so using `percentile_value` might be a better trade-off.


But then the docstring would need a similar update.

Reply:
See https://github.com/scikit-learn/scikit-learn/pull/29034#discussion_r1798986385